### PR TITLE
142 update semantics app

### LINKED ISSUE #142: Update semantics app
Improve semantics app and add new features like new search or context menu for entities

### PR DESCRIPTION
**Pull request by JRadebold**

### CODE REVIEW

**Comment by djs0109** [comment]
@JRadebold @SBlechmann I see that this feature is still not there in this branch https://github.com/N5GEH/n5geh.tools.entirety/pull/105#issuecomment-1695238735 . Is the plan that to implement it in another branch/issue?

File: app/Entirety/entities/views.py [hunk 1]
```diff
@@ -81,21 +84,34 @@ def get_context_data(self, **kwargs):
...
+        
         if self.request.POST.get("Edit"):
+            if not selected:
+                messages.warning(
+                    self.request,
+                    "Please select an entity from the table to edit.",
+                )
+                return redirect("projects:entities:list", project_id=self.project.uuid)
+            
             # TODO: error if more than one selected for edit
```
Comment (djs0109) [inline app/Entirety/entities/views.py]: Is this TODO still active?

File: app/Entirety/entities/views.py [hunk 2]
```diff
@@ -63,6 +65,7 @@ def get_table_data(self):
             messages.error(self.request, e)
             return []
 
+
```
Comment (djs0109) [inline app/Entirety/entities/views.py]: Extra Space

File: app/Entirety/entities/views.py [hunk 3]
```diff
@@ -53,6 +53,8 @@ def get_table_data(self):
         search_option = self.request.GET.get("search-options", default="")
         search_id = ""
         search_type = ""
+        if not search_option and self.request.POST.get("Edit"):
```
Comment (djs0109) [inline app/Entirety/entities/views.py]: Why is this code block needed? In which case will it be activated?

File: app/Entirety/semantics/views.py [hunk 4]
```diff
@@ -15,9 +15,11 @@ class SemanticsVisualizer(ProjectContextMixin, TemplateView):
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
         # PrepData.generate_df(self)
-        context['elements'] = PrepData.generate_df(self)
+        context['elements'] = PrepData.generate_df(self)[0]
         context['types'] = PrepData.types(self)
         context['relationships'] = PrepData.relationships(self)
+        context['entity_ids'] = PrepData.generate_df(self)[1]
+        context['entity_names'] = PrepData.generate_df(self)[2]
```
Comment (djs0109) [inline app/Entirety/semantics/views.py]: Is it necessary to invoke the `generate_df` for three times?

File: app/Entirety/semantics/templates/semantics/semantics_visualize.html [hunk 5]
```diff
@@ -223,10 +241,18 @@ <h3 class="h3">Detail Level</h3>
         relcheckboxes.forEach(function (checkbox) {
             checkbox.addEventListener('change', add_rel_legend)
         });
+        var tests = ['Store', 'Shelf', 'Product']
```
Comment (djs0109) [inline app/Entirety/semantics/templates/semantics/semantics_visualize.html]: Is this list still used somewhere?

File: app/Entirety/entities/views.py (hunk 2 above)
Comment (djs0109) [reply app/Entirety/entities/views.py]: I fixed it

**Comment by djs0109** [comment]
@JRadebold Generally it looks good to me, I have only have some small comments above. Besides, I found slightly different behavior in different browsers (Firefox and Chrome). Could you check it out too?

**Comment by JRadebold** [comment]
> @JRadebold @SBlechmann I see that this feature is still not there in this branch [#105 (comment)](https://github.com/N5GEH/n5geh.tools.entirety/pull/105#issuecomment-1695238735) . Is the plan that to implement it in another branch/issue?

**Comment by JRadebold** [comment]
> @JRadebold Generally it looks good to me, I have only have some small comments above. Besides, I found slightly different behavior in different browsers (Firefox and Chrome). Could you check it out too?

I checkt it out too, but couldnt find any major differences. What differences were there in your case?

**Comment by JRadebold** [comment]
> > @JRadebold @SBlechmann I see that this feature is still not there in this branch [#105 (comment)](https://github.com/N5GEH/n5geh.tools.entirety/pull/105#issuecomment-1695238735) . Is the plan that to implement it in another branch/issue?

Maybe we have misunderstood each other? Now I have solved the problem by adding the dropdown menu with possible matches as soon as you start typing in the search field. I found highliting all entities in real time in the grapf rather confusing, especily with IDs, since usually they all start with "urn:ngsi-ld" and then all entites are higlighted.

File: app/Entirety/entities/views.py (hunk 1 above)
Comment (JRadebold) [reply app/Entirety/entities/views.py]: I am not sure. I checkt it and if you select multiple entites for edit just one entity is randomly selceted. So the app continues working, but still nor error is raised, so i guess this TODO is still active. However this belongs to the "enties" app and does not concern the semantics app. So maybe this is somthing for a new issue?

File: app/Entirety/entities/views.py (hunk 3 above)
Comment (JRadebold) [reply app/Entirety/entities/views.py]: Good question, i have no idea :D

File: app/Entirety/semantics/views.py (hunk 4 above)
Comment (JRadebold) [reply app/Entirety/semantics/views.py]: No its not, good catch. Changed it

File: app/Entirety/semantics/templates/semantics/semantics_visualize.html (hunk 5 above)
Comment (JRadebold) [reply app/Entirety/semantics/templates/semantics/semantics_visualize.html]: No, deleted it

File: app/Entirety/entities/views.py (hunk 3 above)
Comment (djs0109) [reply app/Entirety/entities/views.py]: Ok, I delete it

**Review by djs0109 — APPROVED** [review]

**Pull request merged**